react/slides/react-router-4.markdown

```diff
@@ -0,0 +1,481 @@
+
+```js
+import React from 'react'
+import ReactDOM from 'react-dom'
+import { Provider } from 'react-redux'
+
+import Application from './Application'
+
+import store from './redux-store'
+
+ReactDOM.render(
+  <Provider store={store}>
+    <Application />
+  </Provider>,
+
+  document.getElementById('app')
+)
+```
+
+--
+
+Import `BrowserRouter` from `react-router-dom` package:
+
+```js
+import { BrowserRouter } from 'react-router-dom'
+```
+
+<br />
+
+And wrap your `Application` with imported `BrowserRouter`:
+
+```js
+...
+
+ReactDOM.render(
+  <Provider store={store}>
+    <BrowserRouter>
+      <Application />
+    </BrowserRouter>
+  </Provider>,
+
+  document.getElementById('app')
+)
+
+...
+```
+
+<br />
+
+WARNING! `BrowserRouter` use `History API` under the hood and it won't work with Github Pages. <br />
+If you are going to deploy your application to Github Pages use `HashRouter` instead.
+
+---
+
+Let's imagine we want to display these components:
+
+```js
+const Home  = () => <span>It is home page!</span>  // for /
+const About = () => <span>It is about page!</span> // for /about
+```
+
+<br />
+
+To make it possible we import `Route` from `react-router-dom` package:
+
+```js
+import { Route } from 'react-router-dom'
+```
+
+<br />
+
+And update our `Application` component:
+
+```js
+const Application = () => (
+  <div>
+    <Route path='/'      component={Home}  />
+    <Route path='/about' component={About} />
+  </div>
+)
+```
+
+--
+
+Visit `http://www.example.com/`:
+
+```html
+...
+
+<span>It is home page!</span>
+
+...
+```
+
+<br />
+
+Visit `http://www.example.com/about`:
+
+```html
+...
+
+<span>It is home page!</span> <!-- WTF? It's /about page -->
+<span>It is about page!</span>
+
+...
+```
+
+--
+
+In order to understand what's happened let's think like `React Router 4`.
+
+By default it checks if `path` property of `<Route ... />` equals
+
+to `window.location` or is a sub of `window.location`.
+
+<br />
+
+For example, when visiting `http://www.example.com/one/two/three`:
+
+```js
+<Route path='/one'           ... /> // will be rendered
+<Route path='/one/two'       ... /> // will be rendered
+<Route path='/one/two/three' ... /> // will be rendered
+```
+
+<br />
+
+In order to disable this super feature we can use `exact` flag:
+
+```js
+<Route exact path='/one'           ... /> // will be rendered only on /one
+<Route exact path='/one/two'       ... /> // will be rendered only on /one/two
+<Route exact path='/one/two/three' ... /> // will be rendered only on /one/two/three
+```
+
+---
+
+Cool! Let's create links for our routes!
+
+<br />
+
+First of all, import `Link` from `react-router-dom` package:
+
+```js
+import { Link } from 'react-router-dom'
+```
+
+<br />
+
+And create links for `Home` and `About` routes:
+
+```js
+const Header = () => (
+  <div>
+    <Link to='/'     >Home </Link>
+    <Link to='/about'>About</Link>
+  </div>
+)
+```
+
+<br />
+
+You can also use `replace` flag for your links:
```

Comment:
It's not clear what replace actually does..
how about

You can also use `replace` to replace the current entry in the history stack instead of adding a new one